File: merry360x_mobile/ios-app/Merry360xMobile/ViewModels/BookingViewModel.swift
import Foundation

@MainActor
final class BookingViewModel: ObservableObject {
    @Published var submitting = false
    @Published var statusMessage: String?

    private let service = SupabaseService()

    func submitSampleBooking(propertyId: String) async {
        guard let service else {
            statusMessage = "MobileConfig is not set. Add Supabase URL + anon key first."
            return
        }

        submitting = true
        statusMessage = nil

        let draft = BookingDraft(
            guestId: "replace-with-real-user-id",
            guestName: "Merry Mobile User",
            guestEmail: "mobile@example.com",
            propertyId: propertyId,
            checkIn: "2026-03-15",
            checkOut: "2026-03-17",
            guests: 2,
            totalPrice: 199500,
            currency: "RWF"
        )

        do {
            try await service.submitBooking(draft)
            statusMessage = "Booking request submitted."
        } catch {
            statusMessage = "Booking failed: \(error.localizedDescription)"
        }

        submitting = false
    }

    func startFlutterwavePayment(email: String, amount: Double, currency: String = "RWF") async {
        guard let service else {
            statusMessage = "MobileConfig is not set. Add Supabase URL + anon key first."
            return
        }

        do {
            let response = try await service.createFlutterwavePayment(payload: [
                "email": email,
                "amount": amount,
                "currency": currency
            ])
            if let link = response["paymentLink"] as? String {
                statusMessage = "Payment link created: \(link)"
            } else {
                statusMessage = "Payment initialized."
            }
        } catch {
            statusMessage = "Payment init failed: \(error.localizedDescription)"
        }
    }
}

File: merry360x_mobile/android-app/app/src/main/java/com/merry360x/mobile/viewmodel/BookingViewModel.kt
package com.merry360x.mobile.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.merry360x.mobile.data.BookingDraft
import com.merry360x.mobile.data.SupabaseApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class BookingUiState(
    val submitting: Boolean = false,
    val statusMessage: String? = null,
)

class BookingViewModel(
    private val api: SupabaseApi,
) : ViewModel() {
    private val _state = MutableStateFlow(BookingUiState())
    val state: StateFlow<BookingUiState> = _state.asStateFlow()

    var selectedListingId: String? = null
    var selectedListingTitle: String? = null

    fun submitSampleBooking() {
        viewModelScope.launch {
            _state.value = _state.value.copy(submitting = true, statusMessage = null)

            val result = api.submitBooking(
                BookingDraft(
                    guestId = "replace-with-real-user-id",
                    guestName = "Merry Mobile User",
                    guestEmail = "mobile@example.com",
                    propertyId = selectedListingId ?: "replace-with-real-property-id",
                    checkIn = "2026-03-15",
                    checkOut = "2026-03-17",
                    guests = 2,
                    totalPrice = 199500.0,
                    currency = "RWF",
                )
            )

            _state.value = if (result.isSuccess) {
                BookingUiState(submitting = false, statusMessage = "Booking request submitted.")
            } else {
                BookingUiState(submitting = false, statusMessage = result.exceptionOrNull()?.message ?: "Booking failed")
            }
        }
    }
}
